File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/irpf/ProporcoesIrpf.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf.Irpf;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf.OcorrenciaDeIrpf;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.juros.ApuracaoDeJuros;
import br.jus.trt8.pjecalc.negocio.dominio.ocorrenciaverba.OcorrenciaDeVerba;
import br.jus.trt8.pjecalc.negocio.dominio.pagamento.CreditosDoReclamante;
import br.jus.trt8.pjecalc.negocio.dominio.verbacalculo.VerbaDeCalculo;
import java.io.Serializable;
import java.math.BigDecimal;

public class ProporcoesIrpf
implements Serializable {
    private static final long serialVersionUID = 1L;
    private BigDecimal proporcaoFerias;
    private BigDecimal proporcaoDecimo;
    private BigDecimal proporcaoAnosAnteriores;
    private BigDecimal proporcaoNormal;
    private BigDecimal proporcaoJurosDecimo;
    private BigDecimal proporcaoJurosFerias;
    private BigDecimal proporcaoJurosAnosAnteriores;
    private BigDecimal proporcaoJurosNormal;
    private BigDecimal proporcaoInssDecimo;
    private BigDecimal proporcaoInssFerias;
    private BigDecimal proporcaoInssAnosAnteriores;
    private BigDecimal proporcaoInssNormal;
    private BigDecimal proporcaoPrevidenciaPrivadaDecimo;
    private BigDecimal proporcaoPrevidenciaPrivadaFerias;
    private BigDecimal proporcaoPrevidenciaPrivadaAnosAnteriores;
    private BigDecimal proporcaoPrevidenciaPrivadaNormal;

    public ProporcoesIrpf(CreditosDoReclamante creditosDoReclamante, Irpf irpf) {
        BigDecimal valorPrincipalCorrigido;
        BigDecimal verbaAnosAnteriores = BigDecimal.ZERO;
        BigDecimal verbaIrpfTotalDevido = BigDecimal.ZERO;
        BigDecimal verbaFerias = BigDecimal.ZERO;
        BigDecimal verbaFeriasGozadas = BigDecimal.ZERO;
        BigDecimal verbaDecimoTerceiro = BigDecimal.ZERO;
        BigDecimal verbaDemaisVerbas = BigDecimal.ZERO;
        HelperDate dataLimiteAnosAnteriores = HelperDate.getInstance(irpf.getCalculo().getDataDeLiquidacao());
        dataLimiteAnosAnteriores.setMonth(0);
        dataLimiteAnosAnteriores.setDay(1);
        for (VerbaDeCalculo verba : irpf.getCalculo().getVerbasAtivas()) {
            if (!verba.getIncidenciaIRPF().booleanValue()) continue;
            verbaIrpfTotalDevido = Utils.somar(verbaIrpfTotalDevido, verba.getValorTotalDiferencaCorrigida(), verbaIrpfTotalDevido);
            for (OcorrenciaDeVerba ocorrenciaVerbas : verba.getOcorrenciasAtivas()) {
                BigDecimal base;
                BigDecimal valorOcorrenciaCorrigida = Utils.arredondarValorMonetario(ocorrenciaVerbas.getDiferencaCorrigida());
                if ((Utils.nulo(irpf.getRegimeDeCaixa()) || !irpf.getRegimeDeCaixa().booleanValue()) && HelperDate.dateBefore(ocorrenciaVerbas.getDataInicial(), dataLimiteAnosAnteriores.getDate())) {
                    if (BigDecimal.ZERO.compareTo(valorOcorrenciaCorrigida) == 0 || !Utils.naoNulo(base = ocorrenciaVerbas.getDiferencaCorrigidaParaCalculoDasIncidencias())) continue;
                    verbaAnosAnteriores = verbaAnosAnteriores.add(base, Utils.CONTEXTO_MATEMATICO);
                    continue;
                }
                switch (verba.getCaracteristica()) {
                    case DECIMO_TERCEIRO_SALARIO: {
                        verbaDecimoTerceiro = verbaDecimoTerceiro.add(valorOcorrenciaCorrigida, Utils.CONTEXTO_MATEMATICO);
                        break;
                    }
                    case FERIAS: {
                        base = ocorrenciaVerbas.getDiferencaCorrigidaParaCalculoDasIncidencias();
                        if (!Utils.naoNulo(base)) break;
                        verbaFerias = verbaFerias.add(base, Utils.CONTEXTO_MATEMATICO);
                        verbaFeriasGozadas = verbaFeriasGozadas.add(base, Utils.CONTEXTO_MATEMATICO);
                        break;
                    }
                    case AVISO_PREVIO: 
                    case COMUM: {
                        verbaDemaisVerbas = verbaDemaisVerbas.add(valorOcorrenciaCorrigida, Utils.CONTEXTO_MATEMATICO);
                    }
                }
            }
        }
        if (irpf.getIncidirSobreJurosDeMora().booleanValue()) {
            BigDecimal jurosAnosAnteriores = BigDecimal.ZERO;
            BigDecimal jurosDecimoTerceiro = BigDecimal.ZERO;
            BigDecimal jurosFerias = BigDecimal.ZERO;
            BigDecimal jurosDemaisVerbas = BigDecimal.ZERO;
            for (ApuracaoDeJuros ocorrenciaJuros : irpf.getCalculo().getApuracoesDeJuros()) {
                if (!irpf.getRegimeDeCaixa().booleanValue() && HelperDate.dateBefore(ocorrenciaJuros.getCompetencia(), dataLimiteAnosAnteriores.getDate())) {
                    jurosAnosAnteriores = jurosAnosAnteriores.add(Utils.arredondarValorMonetario(ocorrenciaJuros.getJurosParaIrpfDecimoTerceiro()));
                    jurosAnosAnteriores = jurosAnosAnteriores.add(Utils.arredondarValorMonetario(ocorrenciaJuros.getJurosParaIrpfFerias()));
                    jurosAnosAnteriores = jurosAnosAnteriores.add(Utils.arredondarValorMonetario(ocorrenciaJuros.getJurosParaIrpfDemaisVerbas()));
                    continue;
                }
                jurosDecimoTerceiro = jurosDecimoTerceiro.add(Utils.arredondarValorMonetario(ocorrenciaJuros.getJurosParaIrpfDecimoTerceiro()));
                jurosFerias = jurosFerias.add(Utils.arredondarValorMonetario(ocorrenciaJuros.getJurosParaIrpfFerias()));
                jurosDemaisVerbas = jurosDemaisVerbas.add(Utils.arredondarValorMonetario(ocorrenciaJuros.getJurosParaIrpfDemaisVerbas()));
            }
            BigDecimal juroPrincipal = creditosDoReclamante.getJuroPrincipal();
            if (BigDecimal.ZERO.compareTo(juroPrincipal) != 0) {
                this.proporcaoJurosAnosAnteriores = Utils.dividir(jurosAnosAnteriores, juroPrincipal);
                this.proporcaoJurosDecimo = Utils.dividir(jurosDecimoTerceiro, juroPrincipal);
                this.proporcaoJurosFerias = Utils.dividir(jurosFerias, juroPrincipal);
                this.proporcaoJurosNormal = Utils.dividir(jurosDemaisVerbas, juroPrincipal);
            }
        }
        if (BigDecimal.ZERO.compareTo(valorPrincipalCorrigido = creditosDoReclamante.getValorPrincipal()) != 0) {
            this.proporcaoDecimo = Utils.dividir(verbaDecimoTerceiro, valorPrincipalCorrigido);
            this.proporcaoFerias = Utils.dividir(verbaFerias, valorPrincipalCorrigido);
            this.proporcaoAnosAnteriores = Utils.dividir(verbaAnosAnteriores, valorPrincipalCorrigido);
            this.proporcaoNormal = Utils.dividir(verbaDemaisVerbas, valorPrincipalCorrigido);
        }
        BigDecimal descontoInssTotal = BigDecimal.ZERO;
        BigDecimal descontoInssExclusiva = BigDecimal.ZERO;
        BigDecimal descontoInssEmSeparado = BigDecimal.ZERO;
        BigDecimal descontoInssAnosAnteriores = BigDecimal.ZERO;
        BigDecimal descontoInssNormal = BigDecimal.ZERO;
        BigDecimal descontoPrevidenciaPrivadaTotal = BigDecimal.ZERO;
        BigDecimal descontoPrevidenciaPrivadaExclusiva = BigDecimal.ZERO;
        BigDecimal descontoPrevidenciaPrivadaEmSeparado = BigDecimal.ZERO;
        BigDecimal descontoPrevidenciaPrivadaAnosAnteriores = BigDecimal.ZERO;
        BigDecimal descontoPrevidenciaPrivadaNormal = BigDecimal.ZERO;
        for (OcorrenciaDeIrpf ocorrencia : irpf.getOcorrencias()) {
            descontoInssTotal = Utils.somar(descontoInssTotal, ocorrencia.getValorContribuicaoSocial(), descontoInssTotal);
            descontoPrevidenciaPrivadaTotal = Utils.somar(descontoPrevidenciaPrivadaTotal, ocorrencia.getValorPrevidenciaPrivada(), descontoPrevidenciaPrivadaTotal);
            switch (ocorrencia.getTipo()) {
                case NORMAL: {
                    descontoInssNormal = Utils.somar(descontoInssNormal, ocorrencia.getValorContribuicaoSocial(), descontoInssNormal);
                    descontoPrevidenciaPrivadaNormal = Utils.somar(descontoPrevidenciaPrivadaNormal, ocorrencia.getValorPrevidenciaPrivada(), descontoPrevidenciaPrivadaNormal);
                    break;
                }
                case TRIBUTACAO_EXCLUSIVA: {
                    descontoInssExclusiva = Utils.somar(descontoInssExclusiva, ocorrencia.getValorContribuicaoSocial(), descontoInssExclusiva);
                    descontoPrevidenciaPrivadaExclusiva = Utils.somar(descontoPrevidenciaPrivadaExclusiva, ocorrencia.getValorPrevidenciaPrivada(), descontoPrevidenciaPrivadaExclusiva);
                    break;
                }
                case TRIBUTACAO_EM_SEPARADO: {
                    descontoInssEmSeparado = Utils.somar(descontoInssEmSeparado, ocorrencia.getValorContribuicaoSocial(), descontoInssEmSeparado);
                    descontoPrevidenciaPrivadaEmSeparado = Utils.somar(descontoPrevidenciaPrivadaEmSeparado, ocorrencia.getValorPrevidenciaPrivada(), descontoPrevidenciaPrivadaEmSeparado);
                    break;
                }
                case RRA_ANOS_ANTERIORES: {
                    descontoInssAnosAnteriores = Utils.somar(descontoInssAnosAnteriores, ocorrencia.getValorContribuicaoSocial(), descontoInssAnosAnteriores);
                    descontoPrevidenciaPrivadaAnosAnteriores = Utils.somar(descontoPrevidenciaPrivadaAnosAnteriores, ocorrencia.getValorPrevidenciaPrivada(), descontoPrevidenciaPrivadaAnosAnteriores);
                }
            }
        }
        if (BigDecimal.ZERO.compareTo(descontoInssTotal) != 0) {
            this.proporcaoInssDecimo = Utils.dividir(descontoInssExclusiva, descontoInssTotal);
            this.proporcaoInssFerias = Utils.dividir(descontoInssEmSeparado, descontoInssTotal);
            this.proporcaoInssAnosAnteriores = Utils.dividir(descontoInssAnosAnteriores, descontoInssTotal);
            this.proporcaoInssNormal = Utils.dividir(descontoInssNormal, descontoInssTotal);
        }
        if (BigDecimal.ZERO.compareTo(descontoPrevidenciaPrivadaTotal) != 0) {
            this.proporcaoPrevidenciaPrivadaDecimo = Utils.dividir(descontoPrevidenciaPrivadaExclusiva, descontoPrevidenciaPrivadaTotal);
            this.proporcaoPrevidenciaPrivadaFerias = Utils.dividir(descontoPrevidenciaPrivadaEmSeparado, descontoPrevidenciaPrivadaTotal);
            this.proporcaoPrevidenciaPrivadaAnosAnteriores = Utils.dividir(descontoPrevidenciaPrivadaAnosAnteriores, descontoPrevidenciaPrivadaTotal);
            this.proporcaoPrevidenciaPrivadaNormal = Utils.dividir(descontoPrevidenciaPrivadaNormal, descontoPrevidenciaPrivadaTotal);
        }
    }

    public BigDecimal getProporcaoFerias() {
        return this.proporcaoFerias;
    }

    public BigDecimal getProporcaoDecimo() {
        return this.proporcaoDecimo;
    }

    public BigDecimal getProporcaoNormal() {
        return this.proporcaoNormal;
    }

    public BigDecimal getProporcaoInssDecimo() {
        return this.proporcaoInssDecimo;
    }

    public BigDecimal getProporcaoInssFerias() {
        return this.proporcaoInssFerias;
    }

    public BigDecimal getProporcaoInssAnosAnteriores() {
        return this.proporcaoInssAnosAnteriores;
    }

    public BigDecimal getProporcaoInssNormal() {
        return this.proporcaoInssNormal;
    }

    public BigDecimal getProporcaoPrevidenciaPrivadaDecimo() {
        return this.proporcaoPrevidenciaPrivadaDecimo;
    }

    public BigDecimal getProporcaoPrevidenciaPrivadaFerias() {
        return this.proporcaoPrevidenciaPrivadaFerias;
    }

    public BigDecimal getProporcaoPrevidenciaPrivadaAnosAnteriores() {
        return this.proporcaoPrevidenciaPrivadaAnosAnteriores;
    }

    public BigDecimal getProporcaoPrevidenciaPrivadaNormal() {
        return this.proporcaoPrevidenciaPrivadaNormal;
    }

    public BigDecimal getProporcaoAnosAnteriores() {
        return this.proporcaoAnosAnteriores;
    }

    public BigDecimal getProporcaoJurosDecimo() {
        return this.proporcaoJurosDecimo;
    }

    public BigDecimal getProporcaoJurosFerias() {
        return this.proporcaoJurosFerias;
    }

    public BigDecimal getProporcaoJurosAnosAnteriores() {
        return this.proporcaoJurosAnosAnteriores;
    }

    public BigDecimal getProporcaoJurosNormal() {
        return this.proporcaoJurosNormal;
    }
}

File: src/lib/pjecalc/core/dominio/calculo/irpf/proporcoes-irpf.ts
/**
 * PJe-Calc v2.15.1 — ProporcoesIrpf
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.irpf.ProporcoesIrpf
 *
 * Ref Java: pjecalc-fonte/.../calculo/irpf/ProporcoesIrpf.java (~221 linhas)
 *
 * Calcula proporções utilizadas no rateio do IRPF:
 *   - proporção de cada grupo de verba (13º, férias, anos anteriores, normal)
 *     em relação ao valor principal corrigido do reclamante
 *   - proporção de juros por grupo (se `incidirSobreJurosDeMora`)
 *   - proporção de INSS e Previdência Privada deduzidos por grupo
 *
 * Consumido por `MaquinaDeCalculoDeIrpf` para distribuir valores de pagamento
 * proporcionalmente entre as ocorrências.
 *
 * Os campos são `null` se a base correspondente for zero (paridade com Java).
 */
import Decimal from 'decimal.js';
import { HelperDate } from '../../../base/comum/helper-date';
import { arredondarValorMonetario } from '../../../base/comum/utils';
import { TipoOcorrenciaIrpfEnum } from '../../../constantes/enums';
import type { Irpf } from './irpf';

const ZERO = new Decimal(0);

/**
 * Interface mínima (ducktype) do CreditosDoReclamante (módulo Pagamento, Fase 9).
 */
export interface CreditosDoReclamanteProporcoes {
  getPagoPrincipal?(): Decimal | null;
  getDiferencaPrincipal?(): Decimal | null;
  getValorPrincipal(): Decimal;
  getJuroPrincipal(): Decimal;
}

/** Caracteristica enum (ducktyping) — igual a `CaracteristicaDaVerbaEnum` */
type Caracteristica = string;
const DECIMO_TERCEIRO_SALARIO: Caracteristica = 'DECIMO_TERCEIRO_SALARIO';
const FERIAS: Caracteristica = 'FERIAS';
const AVISO_PREVIO: Caracteristica = 'AVISO_PREVIO';
const COMUM: Caracteristica = 'COMUM';

interface VerbaLike {
  getIncidenciaIRPF?(): boolean;
  getValorTotalDiferencaCorrigida?(): Decimal | null;
  getCaracteristica?(): Caracteristica;
  getOcorrenciasAtivas?(): Iterable<OcorrenciaVerbaLike>;
}

interface OcorrenciaVerbaLike {
  getDiferencaCorrigida?(): Decimal | null;
  getDiferencaCorrigidaParaCalculoDasIncidencias?(): Decimal | null;
  getDataInicial?(): Date | null;
}

interface ApuracaoJurosLike {
  getCompetencia?(): Date | null;
  getJurosParaIrpfDecimoTerceiro?(): Decimal | null;
  getJurosParaIrpfFerias?(): Decimal | null;
  getJurosParaIrpfDemaisVerbas?(): Decimal | null;
}

export class ProporcoesIrpf {
  private proporcaoFerias: Decimal | null = null;
  private proporcaoDecimo: Decimal | null = null;
  private proporcaoAnosAnteriores: Decimal | null = null;
  private proporcaoNormal: Decimal | null = null;

  private proporcaoJurosDecimo: Decimal | null = null;
  private proporcaoJurosFerias: Decimal | null = null;
  private proporcaoJurosAnosAnteriores: Decimal | null = null;
  private proporcaoJurosNormal: Decimal | null = null;

  private proporcaoInssDecimo: Decimal | null = null;
  private proporcaoInssFerias: Decimal | null = null;
  private proporcaoInssAnosAnteriores: Decimal | null = null;
  private proporcaoInssNormal: Decimal | null = null;

  private proporcaoPrevidenciaPrivadaDecimo: Decimal | null = null;
  private proporcaoPrevidenciaPrivadaFerias: Decimal | null = null;
  private proporcaoPrevidenciaPrivadaAnosAnteriores: Decimal | null = null;
  private proporcaoPrevidenciaPrivadaNormal: Decimal | null = null;

  constructor(creditosDoReclamante: CreditosDoReclamanteProporcoes, irpf: Irpf) {
    const calc = irpf.getCalculo();
    if (!calc) return;
    const calcExt = calc as unknown as {
      getDataDeLiquidacao?(): Date;
      getVerbasAtivas?(): Iterable<VerbaLike>;
      getApuracoesDeJuros?(): Iterable<ApuracaoJurosLike>;
    };

    // Data limite "anos anteriores" = 01/01 do ano da liquidação
    const dataLiq = calcExt.getDataDeLiquidacao?.();
    if (!dataLiq) return;
    const dataLimiteAnosAnteriores = new Date(dataLiq.getFullYear(), 0, 1);

    let verbaAnosAnteriores = ZERO;
    let verbaFerias = ZERO;
    let verbaDecimoTerceiro = ZERO;
    let verbaDemaisVerbas = ZERO;

    const regimeDeCaixa = irpf.getRegimeDeCaixa();

    for (const verba of calcExt.getVerbasAtivas?.() ?? []) {
      if (!verba.getIncidenciaIRPF?.()) continue;
      for (const ocVerba of verba.getOcorrenciasAtivas?.() ?? []) {
        const valorOcorrenciaCorrigida = arredondarValorMonetario(ocVerba.getDiferencaCorrigida?.() ?? null);
        const dataInicial = ocVerba.getDataInicial?.();
        const base = ocVerba.getDiferencaCorrigidaParaCalculoDasIncidencias?.() ?? null;

        // Ramo "anos anteriores" (regimeDeCaixa=false e data anterior a 01/01 da liq.)
        if (!regimeDeCaixa && dataInicial && HelperDate.dateBefore(dataInicial, dataLimiteAnosAnteriores)) {
          if (valorOcorrenciaCorrigida.isZero() || base === null) continue;
          verbaAnosAnteriores = verbaAnosAnteriores.plus(base);
          continue;
        }

        switch (verba.getCaracteristica?.()) {
          case DECIMO_TERCEIRO_SALARIO:
            verbaDecimoTerceiro = verbaDecimoTerceiro.plus(valorOcorrenciaCorrigida);
            break;
          case FERIAS:
            if (base !== null) verbaFerias = verbaFerias.plus(base);
            break;
          case AVISO_PREVIO:
          case COMUM:
            verbaDemaisVerbas = verbaDemaisVerbas.plus(valorOcorrenciaCorrigida);
            break;
        }
      }
    }

    // Juros — somente se incidirSobreJurosDeMora
    if (irpf.getIncidirSobreJurosDeMora()) {
      let jurosAnosAnteriores = ZERO;
      let jurosDecimoTerceiro = ZERO;
      let jurosFerias = ZERO;
      let jurosDemaisVerbas = ZERO;
      for (const ocJuros of calcExt.getApuracoesDeJuros?.() ?? []) {
        const comp = ocJuros.getCompetencia?.();
        const valorDec = arredondarValorMonetario(ocJuros.getJurosParaIrpfDecimoTerceiro?.() ?? null);
        const valorFer = arredondarValorMonetario(ocJuros.getJurosParaIrpfFerias?.() ?? null);
        const valorDem = arredondarValorMonetario(ocJuros.getJurosParaIrpfDemaisVerbas?.() ?? null);
        if (!regimeDeCaixa && comp && HelperDate.dateBefore(comp, dataLimiteAnosAnteriores)) {
          jurosAnosAnteriores = jurosAnosAnteriores.plus(valorDec).plus(valorFer).plus(valorDem);
          continue;
        }
        jurosDecimoTerceiro = jurosDecimoTerceiro.plus(valorDec);
        jurosFerias = jurosFerias.plus(valorFer);
        jurosDemaisVerbas = jurosDemaisVerbas.plus(valorDem);
      }
      const juroPrincipal = creditosDoReclamante.getJuroPrincipal();
      if (!juroPrincipal.isZero()) {
        this.proporcaoJurosAnosAnteriores = jurosAnosAnteriores.div(juroPrincipal);
        this.proporcaoJurosDecimo = jurosDecimoTerceiro.div(juroPrincipal);
        this.proporcaoJurosFerias = jurosFerias.div(juroPrincipal);
        this.proporcaoJurosNormal = jurosDemaisVerbas.div(juroPrincipal);
      }
    }

    // Proporções de verba sobre o principal
    const valorPrincipalCorrigido = creditosDoReclamante.getValorPrincipal();
    if (!valorPrincipalCorrigido.isZero()) {
      this.proporcaoDecimo = verbaDecimoTerceiro.div(valorPrincipalCorrigido);
      this.proporcaoFerias = verbaFerias.div(valorPrincipalCorrigido);
      this.proporcaoAnosAnteriores = verbaAnosAnteriores.div(valorPrincipalCorrigido);
      this.proporcaoNormal = verbaDemaisVerbas.div(valorPrincipalCorrigido);
    }

    // INSS / Previdência Privada por tipo de ocorrência
    let descontoInssTotal = ZERO;
    let descontoInssExclusiva = ZERO;
    let descontoInssEmSeparado = ZERO;
    let descontoInssAnosAnteriores = ZERO;
    let descontoInssNormal = ZERO;
    let descontoPPTotal = ZERO;
    let descontoPPExclusiva = ZERO;
    let descontoPPEmSeparado = ZERO;
    let descontoPPAnosAnteriores = ZERO;
    let descontoPPNormal = ZERO;

    for (const oc of irpf.getOcorrencias()) {
      const cs = oc.getValorContribuicaoSocial() ?? ZERO;
      const pp = oc.getValorPrevidenciaPrivada() ?? ZERO;
      descontoInssTotal = descontoInssTotal.plus(cs);
      descontoPPTotal = descontoPPTotal.plus(pp);
      switch (oc.getTipo()) {
        case TipoOcorrenciaIrpfEnum.NORMAL:
          descontoInssNormal = descontoInssNormal.plus(cs);
          descontoPPNormal = descontoPPNormal.plus(pp);
          break;
        case TipoOcorrenciaIrpfEnum.TRIBUTACAO_EXCLUSIVA:
          descontoInssExclusiva = descontoInssExclusiva.plus(cs);
          descontoPPExclusiva = descontoPPExclusiva.plus(pp);
          break;
        case TipoOcorrenciaIrpfEnum.TRIBUTACAO_EM_SEPARADO:
          descontoInssEmSeparado = descontoInssEmSeparado.plus(cs);
          descontoPPEmSeparado = descontoPPEmSeparado.plus(pp);
          break;
        case TipoOcorrenciaIrpfEnum.RRA_ANOS_ANTERIORES:
          descontoInssAnosAnteriores = descontoInssAnosAnteriores.plus(cs);
          descontoPPAnosAnteriores = descontoPPAnosAnteriores.plus(pp);
          break;
      }
    }

    if (!descontoInssTotal.isZero()) {
      this.proporcaoInssDecimo = descontoInssExclusiva.div(descontoInssTotal);
      this.proporcaoInssFerias = descontoInssEmSeparado.div(descontoInssTotal);
      this.proporcaoInssAnosAnteriores = descontoInssAnosAnteriores.div(descontoInssTotal);
      this.proporcaoInssNormal = descontoInssNormal.div(descontoInssTotal);
    }
    if (!descontoPPTotal.isZero()) {
      this.proporcaoPrevidenciaPrivadaDecimo = descontoPPExclusiva.div(descontoPPTotal);
      this.proporcaoPrevidenciaPrivadaFerias = descontoPPEmSeparado.div(descontoPPTotal);
      this.proporcaoPrevidenciaPrivadaAnosAnteriores = descontoPPAnosAnteriores.div(descontoPPTotal);
      this.proporcaoPrevidenciaPrivadaNormal = descontoPPNormal.div(descontoPPTotal);
    }
  }

  getProporcaoFerias(): Decimal | null { return this.proporcaoFerias; }
  getProporcaoDecimo(): Decimal | null { return this.proporcaoDecimo; }
  getProporcaoAnosAnteriores(): Decimal | null { return this.proporcaoAnosAnteriores; }
  getProporcaoNormal(): Decimal | null { return this.proporcaoNormal; }

  getProporcaoJurosDecimo(): Decimal | null { return this.proporcaoJurosDecimo; }
  getProporcaoJurosFerias(): Decimal | null { return this.proporcaoJurosFerias; }
  getProporcaoJurosAnosAnteriores(): Decimal | null { return this.proporcaoJurosAnosAnteriores; }
  getProporcaoJurosNormal(): Decimal | null { return this.proporcaoJurosNormal; }

  getProporcaoInssDecimo(): Decimal | null { return this.proporcaoInssDecimo; }
  getProporcaoInssFerias(): Decimal | null { return this.proporcaoInssFerias; }
  getProporcaoInssAnosAnteriores(): Decimal | null { return this.proporcaoInssAnosAnteriores; }
  getProporcaoInssNormal(): Decimal | null { return this.proporcaoInssNormal; }

  getProporcaoPrevidenciaPrivadaDecimo(): Decimal | null { return this.proporcaoPrevidenciaPrivadaDecimo; }
  getProporcaoPrevidenciaPrivadaFerias(): Decimal | null { return this.proporcaoPrevidenciaPrivadaFerias; }
  getProporcaoPrevidenciaPrivadaAnosAnteriores(): Decimal | null { return this.proporcaoPrevidenciaPrivadaAnosAnteriores; }
  getProporcaoPrevidenciaPrivadaNormal(): Decimal | null { return this.proporcaoPrevidenciaPrivadaNormal; }
}
